### PR TITLE
Samoline expand synth generation

### PR DESCRIPTION
**Pull request by samoline1**
Expanded the synthetic generation process to allow a mix of the previous and the new method described below:

- completion tasks: Uses previous synth generation
- finetuning tasks: Mixes previous approach with new one, which reformulates the output and gets the output, before generating inputs with correctness constraints
- experiments showed it improves robustness to cheating by overfitting test set
- Refactored synth.py for more code modularity

### CODE REVIEW

File: core/models/utility_models.py
Comment (wanderingweights) [inline core/models/utility_models.py:1]: Can we make these more intuitively named, or have a comment - its not super clear

**Review by wanderingweights — CHANGES_REQUESTED** [review]

File: validator/core/cycle.py [hunk 1]
```diff
@@ -43,8 +43,9 @@ async def _run_task_prep(task: Task, keypair: Keypair) -> Task:
         for i in [task.system, task.instruction, task.input, task.output]
         if i is not None
     ]
+    column_to_reformulate = task.output
```
[the remark this answers is not in the document]
Comment (samoline1) [reply validator/core/cycle.py:46]: Sure, I had thought a lot about this
I'll put it inside get_additional_synth_data, and take the last column if size>1 else None
This is less explicit but yields the same result modulo we don't change the order of the columns to sample in the code later on and forgot about this bit

**Comment by samoline1** [comment]
[Linked context](https://discord.com/channels/1272221995400167588/1313119865833525348)

**Review by wanderingweights — APPROVED** [review]
